### PR TITLE
[WIP] Annotation support for Services

### PR DESCRIPTION
**Pull request by weaverryan**
| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | #21103 
| License       | MIT
| Doc PR        | TODO

Hi guys!

This is a WIP solution for adding annotations to services, and is similar to some proposals in #21103. Notice, this does NOT include a "discovery" mechanism for services: you must still register your service like normal (e.g. in YML or XML file), but then the annotations in the service are used to augment the service Definition. But, if #21289 is merged, then that'll cover automatic "discovery" for these services :).

For me, the BIG motivation is to augment autowiring, when you have just one argument that can't be autowired. Annotations read much better than, for example, setting the index "2" argument in YAML.

```yml
_defaults:
    autowire: true
    tags:
        # activates the feature
        - { name: annotated }

parameters:
    mandrill_api_key: XXXXXXXXX

services:
    App\NewsletterManager: ~
        # you could add tags, arguments or anything else here
        # these would override the annotations, if they overlap
```

```php
use Doctrine\ORM\EntityManager;
use Psr\Log\LoggerInterface;

/**
 * You could optionally configure any service definition options here:
 * @DI\Definition(public=false)
 */
class NewsletterManager
{
    private $logger;
    private $em;
    private $mandrillApiKey;
    private $serializer;

    /**
     * @DI\Argument(name="mandrillApiKey", value="%mandrill_api_key%")
     */
    public function __construct(LoggerInterface $logger, EntityManager $em, $mandrillApiKey)
    {
        $this->logger = $logger;
        $this->em = $em;
        $this->mandrillApiKey = $mandrillApiKey;
    }
}
```

Pretty much anything that you can set in YAML would be supported (most is still a TODO). The feature is currently activated by a tag... which is really ugly (though `_defaults` at least helps this). We could add this to the `Definition` object (like we did with `autowire`) if we want.

In #22103, there was some talk of following a Spring standard. While I think it IS important to look and learn from them, this keeps very close to *our* existing standard: effectively re-using all our existing terminology, but in YAML. The `@Argument` annotation re-uses all the options from the `<argument>` tag in XML service config.

Cheers!

TODOS
* Add support for calls
* Add support for factory, deprecated, configurator, decoratedService, autowiringTypes and autowiring/autowiringMethods
* Add support for injection on properties
* Add an `@Tag` annotation
* Finish collection and iterator argument types
* Handle parent class / interface annotations
* Do not throw an exception when you inherit an annotation from a base class, but that arg doesn't exist on the overridden constructor (would currently throw an error)
* Add a new "Resource" class so that the container cache is rebuilt (and rebuilt intelligently)
* Add a lot more tests for many edge cases

### CODE REVIEW

File: src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+     * @var AnnotationReader
+     */
+    private $reader;
+
+    /**
+     * {@inheritdoc}
+     */
+    public function process(ContainerBuilder $container)
+    {
+        if (class_exists(AnnotationReader::class)) {
```
Comment (dosten) [inline src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php]: `if (!class_exists(AnnotationReader::class))` ?

File: src/Symfony/Component/DependencyInjection/Annotation/Service.php [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+/**
+ * @Annotation
+ * @Target({"CLASS"})
+ *
+ * @author Ryan Weaver <ryan@knpuniversity.com>
+ */
+class Service
+{
+    private $shared;
```
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Annotation/Service.php]: Should we add `autowired` too?

File: src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php [hunk 3]
```diff
@@ -0,0 +1,185 @@
...
+        $this->reader = new AnnotationReader();
+
+        $annotatedServiceIds = $container->findTaggedServiceIds('annotated');
+
+        foreach ($annotatedServiceIds as $annotatedServiceId => $params) {
+            $this->augmentServiceDefinition($container->getDefinition($annotatedServiceId));
+        }
+    }
+
+    private function augmentServiceDefinition(Definition $definition)
```
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php]: It should read recursively annotations from parent classes and implemented interfaces.

File: src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+        $argumentIndexes = $this->getMethodArguments($reflectionMethod);
+        foreach ($argAnnotations as $arg) {
+            if (!isset($argumentIndexes[$arg->getName()])) {
+                throw new \InvalidArgumentException(sprintf('Invalid argument name "%s" used on the Argument annotation of %s::%s', $arg->getName(), $definition->getClass(), $reflectionMethod->getName()));
+            }
+            $key = $argumentIndexes[$arg->getName()];
+
+            $onInvalid = $arg->getOnInvalid();
+            $invalidBehavior = ContainerInterface::EXCEPTION_ON_INVALID_REFERENCE;
+            if ('ignore' == $onInvalid) {
```
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php]: `===`

File: src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php [hunk 5]
```diff
@@ -0,0 +1,185 @@
...
+                $invalidBehavior = ContainerInterface::IGNORE_ON_INVALID_REFERENCE;
+            } elseif ('null' == $onInvalid) {
```
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php]: `===`

File: src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+     *
+     * @param \ReflectionMethod $method
+     *
+     * @return array
+     */
+    private function getMethodArguments(\ReflectionMethod $method)
+    {
+        $arguments = array();
+        $i = 0;
+        foreach ($method->getParameters() as $parameter) {
```
Comment (dunglas) [inline src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php]: Why not using the key instead of the `$i` variable?

```php
foreach($method->getParameters() as $i => $parameter) {
    $arguments[$parameter->getName()] = $i;
}
```

It should do the same thing: https://3v4l.org/Zpq4W

**Review by dunglas — CHANGES_REQUESTED** [review]

**Comment by dunglas** [comment]
I'm still not a fond of having (non-standard) annotations for DI. I've explained my reasons in depth here: https://github.com/symfony/symfony/issues/21103#issuecomment-270062867

However I recognize that it is practical for parameters injection. I've proposed a less intrusive alternative in https://github.com/symfony/symfony/pull/20738, but I'm not very convinced by it either.

Maybe should we keep with the current solution for now:

```yaml
_defaults:
    autowire: true

parameters:
    mandrill_api_key: XXXXXXXXX

services:
    App\NewsletterManager: ['', '', '%mandrill_api_key%'] # This is already working
```

**Comment by ogizanagi** [comment]
@dunglas : 

```yaml
services:
    App\NewsletterManager: ['', '', '%mandrill_api_key%'] # This is already working
```

This is not already working, until #21313 is merged 😄

File: src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php (hunk 1 above)
Comment (weaverryan) [inline src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php]: thanks :)

File: src/Symfony/Component/DependencyInjection/Annotation/Service.php (hunk 2 above)
Comment (weaverryan) [inline src/Symfony/Component/DependencyInjection/Annotation/Service.php]: Definitely - it's one of my todos :)

File: src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php (hunk 6 above)
Comment (weaverryan) [inline src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php]: Updated!

**Comment by weaverryan** [comment]
@dunglas I agree with much of what you are saying :), and I see that we're both after trying to solve the "parameter" problem in some decent way. I'm also not convinced by #20738, so we could either stay with the current YAML solution or use annotations (or maybe there's a 3rd option). About your comments (https://github.com/symfony/symfony/issues/21103#issuecomment-270062867), I don't see these as a significant problem. Overall, the fact that this feature would be meant for end-users - and not library authors - puts most of those concerns to rest. And in app code, coupling your code to your framework isn't a problem (actually, I see it as an advantage for speed and clarity). Also, it may *help* the container to be seen as simply an "internal detail" to end-users. If auto-wiring were perfect, then life would be good. But since there will always be edge-cases (parameters being the big one), with annotations, the user could avoid needing to suddenly (and strangely) go into `services.yml` and add a service key + arguments key (e.g. where the first two args are empty strings). Instead, they would just annotate one arg in their class :).

But, we'll see what others think - I wanted to get some real code to help with this exact discussion!

Cheers!

**Comment by GuilhemN** [comment]
> For me, the BIG motivation is to augment autowiring, when you have just one argument that can't be autowired. Annotations read much better than, for example, setting the index "2" argument in YAML.

This made me think, can't we have named arguments in yaml? (e.g.
```yml
services:
    arguments:
        mandrillApiKey: %mandrill_api_key%
```
).

Having annotations look huge to me if this is the only motivation.

**Comment by dunglas** [comment]
@GuilhemN I've discussed exactly with @nicolas-grekas today and I'm working on a PR to implement here 🙌

**Comment by weaverryan** [comment]
@dunglas @GuilhemN I don't think I agree with that approach, as it's unexpected that changing argument names would have a side effect (this is true with annotations too, but at least the annotation & arg name are together). But, let's open that PR and decide the best solution!

**Comment by dunglas** [comment]
Alternative proposal: https://github.com/symfony/symfony/pull/21383

**Comment by nicolas-grekas** [comment]
Note that to me, this PR and #21383 do not compete. We could very well have both onboard.

File: src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+    /**
+     * {@inheritdoc}
+     */
+    public function process(ContainerBuilder $container)
+    {
+        if (!class_exists(AnnotationReader::class)) {
+            return;
+        }
+
+        $this->reader = new AnnotationReader();
```
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php]: what about using `$container->get('annotation_reader')` instead? this service is already used at compile time by some other extensions.

Comment (stof) [inline src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php]: and it already causes issues regarding incomplete service configuration (remember the burden when switching the cache configuration in 3.2 ?). We always documented that getting service instances during a compiler pass is an unsupported usage of the container (and so you are on your own and it may break in weird ways), so we should not do it in Symfony itself (especially when it can still break because of a change done by another bundle)

File: src/Symfony/Component/DependencyInjection/Annotation/Service.php [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+
+    private $public;
+
+    private $synthetic;
+
+    private $abstract;
+
+    private $lazy;
+
+    public function __construct(array $data)
```
Comment (stof) [inline src/Symfony/Component/DependencyInjection/Annotation/Service.php]: why not removing the constructor and using public properties, letting the annotation parser validate properties itself ? This is the recommended usage of doctrine/annotations when annnotation classes don't need to be reused for other purposes (which is why we don't do in the validator component: constraints themselves are used as annotations, but they are not only annotations)

File: src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php (hunk 7 above)
Comment (nicolas-grekas) [inline src/Symfony/Component/DependencyInjection/Compiler/ServiceAnnotationsPass.php]: I more than remember since I worked on that again in #21381
Yet it allowed to fix a core issue, which is that there should be no cache enabled at build time.
Thus, using `annotation_reader` could be also seen as a way for us to support using it as a first class citizen - which we have to anyway.

**Comment by weaverryan** [comment]
I'll close this for now (thanks to #21383). We may still want this in the future, to avoid needing to touch a YAML file once you need to map a custom, non-autowired arg... but let's see how this all sorts out.
